### PR TITLE
Persists latest_gc_cutoff_lsn before performing GC

### PR DESCRIPTION
**Pull request by knizhnik**

### CODE REVIEW

**Comment by knizhnik** [comment]
> Needs a test

I have created the test using fail_points, it actually reproduces "could not find data for key" error but...  this error is just logged and ignored (GC thread is restarted):
```
2022-10-08T05:49:20.924262Z ERROR gc_loop{tenant_id=e75c3d6bde49c666353c3ddae20c43ce}: Gc failed, retrying: could not find data for key 010000000000000000000000000000000000 at LSN 0/8410799, for request at LSN 0/8FC0668: layer traversal: result Continue, cont_lsn 0/8410799, layer: 000000067F000032D2000040140000000A8F-030000000000000000000000000000000002__0000000008410799-0000000009945B29
```
So test is not detected it and considered as successfully completed.
I wonder what can be the best way to catch this error?
- Check presence of particular fail point and if so make this error fatal?
- Grep log for presence of this error?
- Always kill pageserver in debug mode rather than restart thread? (it can be useful not only for GC, but for other backgroud tasks)

File: test_runner/regress/test_gc_cutoff.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    pg_bin.run_capture(["pgbench", "-i", f"-s{scale}", connstr])
+
+    pageserver_http.configure_failpoints(("gc-before-save-metadata", "return"))
+
+    for i in range(5):
+        try:
+            pg_bin.run_capture(["pgbench", "-T100", connstr])
+        except Exception:
+            env.pageserver.stop()
+            env.pageserver.start()
```
Comment (lubennikovaav) [inline test_runner/regress/test_gc_cutoff.py:37]: Don't we need to actually access data to ensure that everything is working fine after failure?

**Review by lubennikovaav — APPROVED** [review]

File: test_runner/regress/test_gc_cutoff.py (hunk 1 above)
Comment (knizhnik) [reply test_runner/regress/test_gc_cutoff.py:37]: Sorry for delay with answer.
For some reasons I was not able to reproduce the problem with this test any more (with most recent main).
May be some change was committed which force flushing metadata file.

In any case - this test is constantly running pgbench, so it does access the data.
But we are catching all exceptions, so we can ignore not onl expected exception caused by fail point, but also other errors.
Let me look at this test more.

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -2076,6 +2077,18 @@ impl Timeline {
...
+            "GC completed removing {} layers, cuttof {}",
+            result.layers_removed, new_gc_cutoff
+        );
+        if result.layers_removed != 0 {
+            fail_point!("gc-before-save-metadata", |_| {
+                info!("Abnormaly terinate pageserver at gc-before-save-metadata fail point");
+                std::process::abort();
+            });
+            return Ok(result);
+        }
```
Comment (hlinnaka) [inline pageserver/src/tenant/timeline.rs:2090]: Hang on, why is the `return Ok(result)` here? So if we decided to remove any layers, we do a quick exit, and skip actually scheduling the deletions in remote storage?

Comment (hlinnaka) [reply pageserver/src/tenant/timeline.rs:2090]: Opened https://github.com/neondatabase/neon/pull/2655 for that

Comment (knizhnik) [reply pageserver/src/tenant/timeline.rs:2090]: Sorry, I dont't remember the reasons of adding this early return.
Looks like it is result of some my experiments when I tried to reproduce the problem.
It is definitely a bug thank you for fixing it.

**Pull request merged**